// @medusajs/medusa/dist/api/routes/admin/shipping-options/create-shipping-option.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminPostShippingOptionsReq = void 0;
var class_transformer_1 = require("class-transformer");
var class_validator_1 = require("class-validator");
var _1 = require(".");
var validator_1 = require("../../../../utils/validator");
/**
 * @oas [post] /shipping-options
 * operationId: "PostShippingOptions"
 * summary: "Create Shipping Option"
 * description: "Creates a Shipping Option"
 * x-authenticated: true
 * requestBody:
 *   content:
 *     application/json:
 *       schema:
 *         properties:
 *           name:
 *             description: "The name of the Shipping Option"
 *             type: string
 *           region_id:
 *             description: "The id of the Region in which the Shipping Option will be available."
 *             type: string
 *           provider_id:
 *             description: "The id of the Fulfillment Provider that handles the Shipping Option."
 *             type: string
 *           profile_id:
 *             description: "The id of the Shipping Profile to add the Shipping Option to."
 *             type: number
 *           data:
 *             description: "The data needed for the Fulfillment Provider to handle shipping with this Shipping Option."
 *             type: object
 *           price_type:
 *             description: "The type of the Shipping Option price."
 *             type: string
 *             enum:
 *               - flat_rate
 *               - calculated
 *           amount:
 *             description: "The amount to charge for the Shipping Option."
 *             type: integer
 *           requirements:
 *             description: "The requirements that must be satisfied for the Shipping Option to be available."
 *             type: array
 *             items:
 *               properties:
 *                 type:
 *                   description: The type of the requirement
 *                   type: string
 *                   enum:
 *                     - max_subtotal
 *                     - min_subtotal
 *                 amount:
 *                   description: The amount to compare with.
 *                   type: integer
 *           is_return:
 *             description: Whether the Shipping Option defines a return shipment.
 *             type: boolean
 *           admin_only:
 *             description: If true, the option can be used for draft orders
 *             type: boolean
 *           metadata:
 *             description: An optional set of key-value pairs with additional information.
 *             type: object
 * tags:
 *   - Shipping Option
 * responses:
 *   200:
 *     description: OK
 *     content:
 *       application/json:
 *         schema:
 *           properties:
 *             shipping_option:
 *               $ref: "#/components/schemas/shipping_option"
 */
exports.default = (function (req, res) { return __awaiter(void 0, void 0, void 0, function () {
    var validated, optionService, shippingProfileService, id, result, data;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, validator_1.validator)(AdminPostShippingOptionsReq, req.body)];
            case 1:
                validated = _a.sent();
                optionService = req.scope.resolve("shippingOptionService");
                shippingProfileService = req.scope.resolve("shippingProfileService");
                if (!!validated.profile_id) return [3 /*break*/, 3];
                return [4 /*yield*/, shippingProfileService.retrieveDefault()];
            case 2:
                id = (_a.sent()).id;
                validated.profile_id = id;
                _a.label = 3;
            case 3: return [4 /*yield*/, optionService.create(validated)];
            case 4:
                result = _a.sent();
                return [4 /*yield*/, optionService.retrieve(result.id, {
                        select: _1.defaultFields,
                        relations: _1.defaultRelations,
                    })];
            case 5:
                data = _a.sent();
                res.status(200).json({ shipping_option: data });
                return [2 /*return*/];
        }
    });
}); });
var OptionRequirement = /** @class */ (function () {
    function OptionRequirement() {
    }
    __decorate([
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], OptionRequirement.prototype, "type", void 0);
    __decorate([
        (0, class_validator_1.IsNumber)(),
        __metadata("design:type", Number)
    ], OptionRequirement.prototype, "amount", void 0);
    return OptionRequirement;
}());
var AdminPostShippingOptionsReq = /** @class */ (function () {
    function AdminPostShippingOptionsReq() {
        this.admin_only = false;
        this.is_return = false;
    }
    __decorate([
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], AdminPostShippingOptionsReq.prototype, "name", void 0);
    __decorate([
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], AdminPostShippingOptionsReq.prototype, "region_id", void 0);
    __decorate([
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], AdminPostShippingOptionsReq.prototype, "provider_id", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], AdminPostShippingOptionsReq.prototype, "profile_id", void 0);
    __decorate([
        (0, class_validator_1.IsObject)(),
        __metadata("design:type", Object)
    ], AdminPostShippingOptionsReq.prototype, "data", void 0);
    __decorate([
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], AdminPostShippingOptionsReq.prototype, "price_type", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsNumber)(),
        __metadata("design:type", Number)
    ], AdminPostShippingOptionsReq.prototype, "amount", void 0);
    __decorate([
        (0, class_validator_1.IsArray)(),
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.ValidateNested)({ each: true }),
        (0, class_transformer_1.Type)(function () { return OptionRequirement; }),
        __metadata("design:type", Array)
    ], AdminPostShippingOptionsReq.prototype, "requirements", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_transformer_1.Type)(function () { return Boolean; }),
        __metadata("design:type", Boolean)
    ], AdminPostShippingOptionsReq.prototype, "admin_only", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_transformer_1.Type)(function () { return Boolean; }),
        __metadata("design:type", Boolean)
    ], AdminPostShippingOptionsReq.prototype, "is_return", void 0);
    __decorate([
        (0, class_validator_1.IsObject)(),
        (0, class_validator_1.IsOptional)(),
        __metadata("design:type", Object)
    ], AdminPostShippingOptionsReq.prototype, "metadata", void 0);
    return AdminPostShippingOptionsReq;
}());
exports.AdminPostShippingOptionsReq = AdminPostShippingOptionsReq;
