// @medusajs/medusa/dist/api/routes/store/customers/list-orders.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoreGetCustomersCustomerOrdersParams = void 0;
var class_transformer_1 = require("class-transformer");
var class_validator_1 = require("class-validator");
var validator_1 = require("../../../../utils/validator");
var orders_1 = require("../orders");
/**
 * @oas [get] /customers/me/orders
 * operationId: GetCustomersCustomerOrders
 * summary: Retrieve Customer Orders
 * description: "Retrieves a list of a Customer's Orders."
 * x-authenticated: true
 * parameters:
 *   - (query) limit {integer} How many addresses to return.
 *   - (query) offset {integer} The offset in the resulting addresses.
 *   - (query) fields {string} (Comma separated string) Which fields should be included in the resulting addresses.
 *   - (query) expand {string} (Comma separated string) Which relations should be expanded in the resulting addresses.
 * tags:
 *   - Customer
 * responses:
 *   200:
 *     description: OK
 *     content:
 *       application/json:
 *         schema:
 *           properties:
 *             count:
 *               description: The total number of Orders.
 *               type: integer
 *             offset:
 *               description: The offset for pagination.
 *               type: integer
 *             limit:
 *               description: The maxmimum number of Orders to return,
 *               type: integer
 *             orders:
 *               type: array
 *               items:
 *                 $ref: "#/components/schemas/orders"
 */
exports.default = (function (req, res) { return __awaiter(void 0, void 0, void 0, function () {
    var id, orderService, selector, validated, includeFields, expandFields, listConfig, _a, orders, count;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                id = req.user.customer_id;
                orderService = req.scope.resolve("orderService");
                selector = {
                    customer_id: id,
                };
                return [4 /*yield*/, (0, validator_1.validator)(StoreGetCustomersCustomerOrdersParams, req.query)];
            case 1:
                validated = _b.sent();
                includeFields = [];
                if (validated.fields) {
                    includeFields = validated.fields.split(",");
                    includeFields = includeFields.filter(function (f) {
                        return orders_1.allowedStoreOrdersFields.includes(f);
                    });
                }
                expandFields = [];
                if (validated.expand) {
                    expandFields = validated.expand.split(",");
                    expandFields = expandFields.filter(function (f) {
                        return orders_1.allowedStoreOrdersRelations.includes(f);
                    });
                }
                listConfig = {
                    select: includeFields.length ? includeFields : orders_1.allowedStoreOrdersFields,
                    relations: expandFields.length ? expandFields : orders_1.allowedStoreOrdersRelations,
                    skip: validated.offset,
                    take: validated.limit,
                    order: { created_at: "DESC" },
                };
                return [4 /*yield*/, orderService.listAndCount(selector, listConfig)];
            case 2:
                _a = __read.apply(void 0, [_b.sent(), 2]), orders = _a[0], count = _a[1];
                res.json({ orders: orders, count: count, offset: validated.offset, limit: validated.limit });
                return [2 /*return*/];
        }
    });
}); });
var StoreGetCustomersCustomerOrdersParams = /** @class */ (function () {
    function StoreGetCustomersCustomerOrdersParams() {
        this.limit = 10;
        this.offset = 0;
    }
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsNumber)(),
        (0, class_transformer_1.Type)(function () { return Number; }),
        __metadata("design:type", Object)
    ], StoreGetCustomersCustomerOrdersParams.prototype, "limit", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsNumber)(),
        (0, class_transformer_1.Type)(function () { return Number; }),
        __metadata("design:type", Object)
    ], StoreGetCustomersCustomerOrdersParams.prototype, "offset", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], StoreGetCustomersCustomerOrdersParams.prototype, "fields", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], StoreGetCustomersCustomerOrdersParams.prototype, "expand", void 0);
    return StoreGetCustomersCustomerOrdersParams;
}());
exports.StoreGetCustomersCustomerOrdersParams = StoreGetCustomersCustomerOrdersParams;
