// @medusajs/medusa/dist/config/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var dotenv_1 = __importDefault(require("dotenv"));
// Set the NODE_ENV to 'development' by default
process.env.NODE_ENV = process.env.NODE_ENV || "development";
var envFound = dotenv_1.default.config();
if (!envFound) {
    // This error should crash whole process
    throw new Error("⚠️  Couldn't find .env file  ⚠️");
}
var config = {
    /**
     * Your favorite port
     */
    port: parseInt(process.env.PORT, 10),
    databaseURL: process.env.MONGODB_URI,
    redisURI: process.env.REDIS_URI,
    /**
     * Your secret sauce
     */
    jwtSecret: process.env.NODE_ENV === "test" ? "test" : process.env.JWT_SECRET,
    cookieSecret: process.env.NODE_ENV === "test" ? "test" : process.env.COOKIE_SECRET,
    /**
     * Used by winston logger
     */
    logs: {
        level: process.env.LOG_LEVEL || "silly",
    },
    /**
     * API configs
     */
    api: {
        prefix: "/api",
    },
};
exports.default = config;
