// @medusajs/medusa/dist/services/event-bus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bull_1 = __importDefault(require("bull"));
var ioredis_1 = __importDefault(require("ioredis"));
/**
 * Can keep track of multiple subscribers to different events and run the
 * subscribers when events happen. Events will run asynchronously.
 * @class
 */
var EventBusService = /** @class */ (function () {
    function EventBusService(_a, config, singleton) {
        var _this = this;
        var manager = _a.manager, logger = _a.logger, stagedJobRepository = _a.stagedJobRepository, redisClient = _a.redisClient, redisSubscriber = _a.redisSubscriber;
        if (singleton === void 0) { singleton = true; }
        /**
         * Handles incoming jobs.
         * @param {Object} job The job object
         * @return {Promise} resolves to the results of the subscriber calls.
         */
        this.worker_ = function (job) {
            var _a = job.data, eventName = _a.eventName, data = _a.data;
            var eventObservers = _this.observers_[eventName] || [];
            var wildcardObservers = _this.observers_["*"] || [];
            var observers = eventObservers.concat(wildcardObservers);
            _this.logger_.info("Processing ".concat(eventName, " which has ").concat(eventObservers.length, " subscribers"));
            return Promise.all(observers.map(function (subscriber) {
                return subscriber(data, eventName).catch(function (err) {
                    _this.logger_.warn("An error occured while processing ".concat(eventName, ": ").concat(err));
                    console.log(err);
                    return err;
                });
            }));
        };
        /**
         * Handles incoming jobs.
         * @param {Object} job The job object
         * @return {Promise} resolves to the results of the subscriber calls.
         */
        this.cronWorker_ = function (job) {
            var _a = job.data, eventName = _a.eventName, data = _a.data;
            var observers = _this.cronHandlers_[eventName] || [];
            _this.logger_.info("Processing cron job: ".concat(eventName));
            return Promise.all(observers.map(function (subscriber) {
                return subscriber(data, eventName).catch(function (err) {
                    _this.logger_.warn("An error occured while processing ".concat(eventName, ": ").concat(err));
                    return err;
                });
            }));
        };
        var opts = {
            createClient: function (type) {
                switch (type) {
                    case "client":
                        return redisClient;
                    case "subscriber":
                        return redisSubscriber;
                    default:
                        if (config.projectConfig.redis_url) {
                            return new ioredis_1.default(config.projectConfig.redis_url);
                        }
                        return redisClient;
                }
            },
        };
        this.config_ = config;
        /** @private {EntityManager} */
        this.manager_ = manager;
        /** @private {logger} */
        this.logger_ = logger;
        this.stagedJobRepository_ = stagedJobRepository;
        if (singleton) {
            /** @private {object} */
            this.observers_ = {};
            /** @private {BullQueue} */
            this.queue_ = new bull_1.default("".concat(this.constructor.name, ":queue"), opts);
            /** @private {object} to handle cron jobs */
            this.cronHandlers_ = {};
            this.redisClient_ = redisClient;
            this.redisSubscriber_ = redisSubscriber;
            /** @private {BullQueue} used for cron jobs */
            this.cronQueue_ = new bull_1.default("cron-jobs:queue", opts);
            // Register our worker to handle emit calls
            this.queue_.process(this.worker_);
            // Register cron worker
            this.cronQueue_.process(this.cronWorker_);
            if (process.env.NODE_ENV !== "test") {
                this.startEnqueuer();
            }
        }
    }
    EventBusService.prototype.withTransaction = function (transactionManager) {
        if (!transactionManager) {
            return this;
        }
        var cloned = new EventBusService({
            manager: transactionManager,
            stagedJobRepository: this.stagedJobRepository_,
            logger: this.logger_,
            redisClient: this.redisClient_,
            redisSubscriber: this.redisSubscriber_,
        }, this.config_, false);
        cloned.transactionManager_ = transactionManager;
        cloned.queue_ = this.queue_;
        return cloned;
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param {string} event - the event that the subscriber will listen for.
     * @param {func} subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     */
    EventBusService.prototype.subscribe = function (event, subscriber) {
        if (typeof subscriber !== "function") {
            throw new Error("Subscriber must be a function");
        }
        if (this.observers_[event]) {
            this.observers_[event].push(subscriber);
        }
        else {
            this.observers_[event] = [subscriber];
        }
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param {string} event - the event that the subscriber will listen for.
     * @param {func} subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     */
    EventBusService.prototype.unsubscribe = function (event, subscriber) {
        if (typeof subscriber !== "function") {
            throw new Error("Subscriber must be a function");
        }
        if (this.observers_[event]) {
            var index = this.observers_[event].indexOf(subscriber);
            if (index !== -1) {
                this.observers_[event].splice(index, 1);
            }
        }
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param {string} event - the event that the subscriber will listen for.
     * @param {func} subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     */
    EventBusService.prototype.registerCronHandler_ = function (event, subscriber) {
        if (typeof subscriber !== "function") {
            throw new Error("Handler must be a function");
        }
        if (this.observers_[event]) {
            this.cronHandlers_[event].push(subscriber);
        }
        else {
            this.cronHandlers_[event] = [subscriber];
        }
    };
    /**
     * Calls all subscribers when an event occurs.
     * @param {string} eventName - the name of the event to be process.
     * @param {?any} data - the data to send to the subscriber.
     * @param {?any} options - options to add the job with
     * @return {BullJob} - the job from our queue
     */
    EventBusService.prototype.emit = function (eventName, data, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var stagedJobRepository, created, opts;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.transactionManager_) return [3 /*break*/, 2];
                        stagedJobRepository = this.transactionManager_.getCustomRepository(this.stagedJobRepository_);
                        return [4 /*yield*/, stagedJobRepository.create({
                                event_name: eventName,
                                data: data,
                            })];
                    case 1:
                        created = _a.sent();
                        return [2 /*return*/, stagedJobRepository.save(created)];
                    case 2:
                        opts = { removeOnComplete: true };
                        if (typeof options.delay === "number") {
                            opts.delay = options.delay;
                        }
                        this.queue_.add({ eventName: eventName, data: data }, opts);
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EventBusService.prototype.sleep = function (ms) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve) {
                        setTimeout(resolve, ms);
                    })];
            });
        });
    };
    EventBusService.prototype.startEnqueuer = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.enRun_ = true;
                this.enqueue_ = this.enqueuer_();
                return [2 /*return*/];
            });
        });
    };
    EventBusService.prototype.stopEnqueuer = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.enRun_ = false;
                        return [4 /*yield*/, this.enqueue_];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    EventBusService.prototype.enqueuer_ = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _loop_1, this_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _loop_1 = function () {
                            var listConfig, sjRepo, jobs;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        listConfig = {
                                            relations: [],
                                            skip: 0,
                                            take: 1000,
                                        };
                                        sjRepo = this_1.manager_.getCustomRepository(this_1.stagedJobRepository_);
                                        return [4 /*yield*/, sjRepo.find({}, listConfig)];
                                    case 1:
                                        jobs = _b.sent();
                                        return [4 /*yield*/, Promise.all(jobs.map(function (job) {
                                                _this.queue_
                                                    .add({ eventName: job.event_name, data: job.data }, { removeOnComplete: true })
                                                    .then(function () { return __awaiter(_this, void 0, void 0, function () {
                                                    return __generator(this, function (_a) {
                                                        switch (_a.label) {
                                                            case 0: return [4 /*yield*/, sjRepo.remove(job)];
                                                            case 1:
                                                                _a.sent();
                                                                return [2 /*return*/];
                                                        }
                                                    });
                                                }); });
                                            }))];
                                    case 2:
                                        _b.sent();
                                        return [4 /*yield*/, this_1.sleep(3000)];
                                    case 3:
                                        _b.sent();
                                        return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        _a.label = 1;
                    case 1:
                        if (!this.enRun_) return [3 /*break*/, 3];
                        return [5 /*yield**/, _loop_1()];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 1];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Registers a cron job.
     * @param {string} eventName - the name of the event
     * @param {object} data - the data to be sent with the event
     * @param {string} cron - the cron pattern
     * @param {function} handler - the handler to call on each cron job
     * @return {void}
     */
    EventBusService.prototype.createCronJob = function (eventName, data, cron, handler) {
        this.logger_.info("Registering ".concat(eventName));
        this.registerCronHandler_(eventName, handler);
        return this.cronQueue_.add({
            eventName: eventName,
            data: data,
        }, { repeat: { cron: cron } });
    };
    return EventBusService;
}());
exports.default = EventBusService;
