// @medusajs/medusa/dist/services/shipping-profile.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = __importDefault(require("lodash"));
var medusa_core_utils_1 = require("medusa-core-utils");
var medusa_interfaces_1 = require("medusa-interfaces");
var typeorm_1 = require("typeorm");
/**
 * Provides layer to manipulate profiles.
 * @class
 * @implements {BaseService}
 */
var ShippingProfileService = /** @class */ (function (_super) {
    __extends(ShippingProfileService, _super);
    function ShippingProfileService(_a) {
        var manager = _a.manager, shippingProfileRepository = _a.shippingProfileRepository, productService = _a.productService, productRepository = _a.productRepository, shippingOptionService = _a.shippingOptionService, customShippingOptionService = _a.customShippingOptionService;
        var _this = _super.call(this) || this;
        /** @private @const {EntityManager} */
        _this.manager_ = manager;
        /** @private @const {ShippingProfileRepository} */
        _this.shippingProfileRepository_ = shippingProfileRepository;
        /** @private @const {ProductService} */
        _this.productService_ = productService;
        /** @private @const {ProductReppsitory} */
        _this.productRepository_ = productRepository;
        /** @private @const {ShippingOptionService} */
        _this.shippingOptionService_ = shippingOptionService;
        /** @private @const {CustomShippingOptionService} */
        _this.customShippingOptionService_ = customShippingOptionService;
        return _this;
    }
    ShippingProfileService.prototype.withTransaction = function (transactionManager) {
        if (!transactionManager) {
            return this;
        }
        var cloned = new ShippingProfileService({
            manager: transactionManager,
            shippingProfileRepository: this.shippingProfileRepository_,
            productService: this.productService_,
            shippingOptionService: this.shippingOptionService_,
            customShippingOptionService: this.customShippingOptionService_,
        });
        cloned.transactionManager_ = transactionManager;
        return cloned;
    };
    /**
     * @param {Object} selector - the query object for find
     * @param {Object} config - the config object for find
     * @return {Promise} the result of the find operation
     */
    ShippingProfileService.prototype.list = function (selector, config) {
        if (selector === void 0) { selector = {}; }
        if (config === void 0) { config = { relations: [], skip: 0, take: 10 }; }
        return __awaiter(this, void 0, void 0, function () {
            var shippingProfileRepo, query;
            return __generator(this, function (_a) {
                shippingProfileRepo = this.manager_.getCustomRepository(this.shippingProfileRepository_);
                query = this.buildQuery_(selector, config);
                return [2 /*return*/, shippingProfileRepo.find(query)];
            });
        });
    };
    ShippingProfileService.prototype.fetchOptionsByProductIds = function (productIds, filter) {
        return __awaiter(this, void 0, void 0, function () {
            var products, profiles, optionIds, options;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.productService_.list({
                            id: (0, typeorm_1.Any)(productIds),
                        }, {
                            relations: [
                                "profile",
                                "profile.shipping_options",
                                "profile.shipping_options.requirements",
                            ],
                        })];
                    case 1:
                        products = _a.sent();
                        profiles = products.map(function (p) { return p.profile; });
                        optionIds = profiles.reduce(function (acc, next) { return acc.concat(next.shipping_options); }, []);
                        return [4 /*yield*/, Promise.all(optionIds.map(function (option) { return __awaiter(_this, void 0, void 0, function () {
                                var canSend;
                                return __generator(this, function (_a) {
                                    canSend = true;
                                    if (filter.region_id) {
                                        if (filter.region_id !== option.region_id) {
                                            canSend = false;
                                        }
                                    }
                                    if (option.deleted_at !== null) {
                                        canSend = false;
                                    }
                                    return [2 /*return*/, canSend ? option : null];
                                });
                            }); }))];
                    case 2:
                        options = _a.sent();
                        return [2 /*return*/, options.filter(function (o) { return !!o; })];
                }
            });
        });
    };
    /**
     * Gets a profile by id.
     * Throws in case of DB Error and if profile was not found.
     * @param {string} profileId - the id of the profile to get.
     * @param {Object} options - options opf the query.
     * @return {Promise<Product>} the profile document.
     */
    ShippingProfileService.prototype.retrieve = function (profileId, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var profileRepository, validatedId, query, profile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        profileRepository = this.manager_.getCustomRepository(this.shippingProfileRepository_);
                        validatedId = this.validateId_(profileId);
                        query = {
                            where: { id: validatedId },
                        };
                        if (options.select) {
                            query.select = options.select;
                        }
                        if (options.relations) {
                            query.relations = options.relations;
                        }
                        return [4 /*yield*/, profileRepository.findOne(query)];
                    case 1:
                        profile = _a.sent();
                        if (!profile) {
                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Profile with id: ".concat(profileId, " was not found"));
                        }
                        return [2 /*return*/, profile];
                }
            });
        });
    };
    ShippingProfileService.prototype.retrieveDefault = function () {
        return __awaiter(this, void 0, void 0, function () {
            var profileRepository, profile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        profileRepository = this.manager_.getCustomRepository(this.shippingProfileRepository_);
                        return [4 /*yield*/, profileRepository.findOne({
                                where: { type: "default" },
                            })];
                    case 1:
                        profile = _a.sent();
                        return [2 /*return*/, profile];
                }
            });
        });
    };
    /**
     * Creates a default shipping profile, if this does not already exist.
     * @return {Promise<ShippingProfile>} the shipping profile
     */
    ShippingProfileService.prototype.createDefault = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var profile, profileRepository, p;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.retrieveDefault()];
                                case 1:
                                    profile = _a.sent();
                                    if (!!profile) return [3 /*break*/, 4];
                                    profileRepository = manager.getCustomRepository(this.shippingProfileRepository_);
                                    return [4 /*yield*/, profileRepository.create({
                                            type: "default",
                                            name: "Default Shipping Profile",
                                        })];
                                case 2:
                                    p = _a.sent();
                                    return [4 /*yield*/, profileRepository.save(p)];
                                case 3:
                                    profile = _a.sent();
                                    _a.label = 4;
                                case 4: return [2 /*return*/, profile];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Retrieves the default gift card profile
     * @return {Object} the shipping profile for gift cards
     */
    ShippingProfileService.prototype.retrieveGiftCardDefault = function () {
        return __awaiter(this, void 0, void 0, function () {
            var profileRepository, giftCardProfile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        profileRepository = this.manager_.getCustomRepository(this.shippingProfileRepository_);
                        return [4 /*yield*/, profileRepository.findOne({
                                where: { type: "gift_card" },
                            })];
                    case 1:
                        giftCardProfile = _a.sent();
                        return [2 /*return*/, giftCardProfile];
                }
            });
        });
    };
    /**
     * Creates a default shipping profile, for gift cards if unless it already
     * exists.
     * @return {Promise<ShippingProfile>} the shipping profile
     */
    ShippingProfileService.prototype.createGiftCardDefault = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var profile, profileRepository, p;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.retrieveGiftCardDefault()];
                                case 1:
                                    profile = _a.sent();
                                    if (!!profile) return [3 /*break*/, 4];
                                    profileRepository = manager.getCustomRepository(this.shippingProfileRepository_);
                                    return [4 /*yield*/, profileRepository.create({
                                            type: "gift_card",
                                            name: "Gift Card Profile",
                                        })];
                                case 2:
                                    p = _a.sent();
                                    return [4 /*yield*/, profileRepository.save(p)];
                                case 3:
                                    profile = _a.sent();
                                    _a.label = 4;
                                case 4: return [2 /*return*/, profile];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Creates a new shipping profile.
     * @param {ShippingProfile} profile - the shipping profile to create from
     * @return {Promise} the result of the create operation
     */
    ShippingProfileService.prototype.create = function (profile) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var profileRepository, created, result;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    profileRepository = manager.getCustomRepository(this.shippingProfileRepository_);
                                    if (profile.products || profile.shipping_options) {
                                        throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "Please add products and shipping_options after creating Shipping Profiles");
                                    }
                                    created = profileRepository.create(profile);
                                    return [4 /*yield*/, profileRepository.save(created)];
                                case 1:
                                    result = _a.sent();
                                    return [2 /*return*/, result];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Updates a profile. Metadata updates and product updates should use
     * dedicated methods, e.g. `setMetadata`, `addProduct`, etc. The function
     * will throw errors if metadata or product updates are attempted.
     * @param {string} profileId - the id of the profile. Must be a string that
     *   can be casted to an ObjectId
     * @param {object} update - an object with the update values.
     * @return {Promise} resolves to the update result.
     */
    ShippingProfileService.prototype.update = function (profileId, update) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var profileRepository, profile, metadata, products, shipping_options, rest, products_1, products_1_1, pId, e_1_1, shipping_options_1, shipping_options_1_1, oId, e_2_1, _a, _b, _c, key, value, result;
                        var e_1, _d, e_2, _e, e_3, _f;
                        return __generator(this, function (_g) {
                            switch (_g.label) {
                                case 0:
                                    profileRepository = manager.getCustomRepository(this.shippingProfileRepository_);
                                    return [4 /*yield*/, this.retrieve(profileId, {
                                            relations: [
                                                "products",
                                                "products.profile",
                                                "shipping_options",
                                                "shipping_options.profile",
                                            ],
                                        })];
                                case 1:
                                    profile = _g.sent();
                                    metadata = update.metadata, products = update.products, shipping_options = update.shipping_options, rest = __rest(update, ["metadata", "products", "shipping_options"]);
                                    if (metadata) {
                                        profile.metadata = this.setMetadata_(profile, metadata);
                                    }
                                    if (!products) return [3 /*break*/, 9];
                                    _g.label = 2;
                                case 2:
                                    _g.trys.push([2, 7, 8, 9]);
                                    products_1 = __values(products), products_1_1 = products_1.next();
                                    _g.label = 3;
                                case 3:
                                    if (!!products_1_1.done) return [3 /*break*/, 6];
                                    pId = products_1_1.value;
                                    return [4 /*yield*/, this.productService_.withTransaction(manager).update(pId, {
                                            profile_id: profile.id,
                                        })];
                                case 4:
                                    _g.sent();
                                    _g.label = 5;
                                case 5:
                                    products_1_1 = products_1.next();
                                    return [3 /*break*/, 3];
                                case 6: return [3 /*break*/, 9];
                                case 7:
                                    e_1_1 = _g.sent();
                                    e_1 = { error: e_1_1 };
                                    return [3 /*break*/, 9];
                                case 8:
                                    try {
                                        if (products_1_1 && !products_1_1.done && (_d = products_1.return)) _d.call(products_1);
                                    }
                                    finally { if (e_1) throw e_1.error; }
                                    return [7 /*endfinally*/];
                                case 9:
                                    if (!shipping_options) return [3 /*break*/, 17];
                                    _g.label = 10;
                                case 10:
                                    _g.trys.push([10, 15, 16, 17]);
                                    shipping_options_1 = __values(shipping_options), shipping_options_1_1 = shipping_options_1.next();
                                    _g.label = 11;
                                case 11:
                                    if (!!shipping_options_1_1.done) return [3 /*break*/, 14];
                                    oId = shipping_options_1_1.value;
                                    return [4 /*yield*/, this.shippingOptionService_
                                            .withTransaction(manager)
                                            .update(oId, {
                                            profile_id: profile.id,
                                        })];
                                case 12:
                                    _g.sent();
                                    _g.label = 13;
                                case 13:
                                    shipping_options_1_1 = shipping_options_1.next();
                                    return [3 /*break*/, 11];
                                case 14: return [3 /*break*/, 17];
                                case 15:
                                    e_2_1 = _g.sent();
                                    e_2 = { error: e_2_1 };
                                    return [3 /*break*/, 17];
                                case 16:
                                    try {
                                        if (shipping_options_1_1 && !shipping_options_1_1.done && (_e = shipping_options_1.return)) _e.call(shipping_options_1);
                                    }
                                    finally { if (e_2) throw e_2.error; }
                                    return [7 /*endfinally*/];
                                case 17:
                                    try {
                                        for (_a = __values(Object.entries(rest)), _b = _a.next(); !_b.done; _b = _a.next()) {
                                            _c = __read(_b.value, 2), key = _c[0], value = _c[1];
                                            profile[key] = value;
                                        }
                                    }
                                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                                    finally {
                                        try {
                                            if (_b && !_b.done && (_f = _a.return)) _f.call(_a);
                                        }
                                        finally { if (e_3) throw e_3.error; }
                                    }
                                    return [4 /*yield*/, profileRepository.save(profile)];
                                case 18:
                                    result = _g.sent();
                                    return [2 /*return*/, result];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Deletes a profile with a given profile id.
     * @param {string} profileId - the id of the profile to delete. Must be
     *   castable as an ObjectId
     * @return {Promise} the result of the delete operation.
     */
    ShippingProfileService.prototype.delete = function (profileId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var profileRepo, profile;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    profileRepo = manager.getCustomRepository(this.shippingProfileRepository_);
                                    return [4 /*yield*/, profileRepo.findOne({ where: { id: profileId } })];
                                case 1:
                                    profile = _a.sent();
                                    if (!profile) {
                                        return [2 /*return*/, Promise.resolve()];
                                    }
                                    return [4 /*yield*/, profileRepo.softRemove(profile)];
                                case 2:
                                    _a.sent();
                                    return [2 /*return*/, Promise.resolve()];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Adds a product to a profile. The method is idempotent, so multiple calls
     * with the same product variant will have the same result.
     * @param {string} profileId - the profile to add the product to.
     * @param {string} productId - the product to add.
     * @return {Promise} the result of update
     */
    ShippingProfileService.prototype.addProduct = function (profileId, productId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var updated;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.productService_
                                        .withTransaction(manager)
                                        .update(productId, { profile_id: profileId })];
                                case 1:
                                    _a.sent();
                                    return [4 /*yield*/, this.retrieve(profileId)];
                                case 2:
                                    updated = _a.sent();
                                    return [2 /*return*/, updated];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Adds a shipping option to the profile. The shipping option can be used to
     * fulfill the products in the products field.
     * @param {string} profileId - the profile to apply the shipping option to
     * @param {string} optionId - the option to add to the profile
     * @return {Promise} the result of the model update operation
     */
    ShippingProfileService.prototype.addShippingOption = function (profileId, optionId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var updated;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.shippingOptionService_
                                        .withTransaction(manager)
                                        .update(optionId, { profile_id: profileId })];
                                case 1:
                                    _a.sent();
                                    return [4 /*yield*/, this.retrieve(profileId)];
                                case 2:
                                    updated = _a.sent();
                                    return [2 /*return*/, updated];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Decorates a profile.
     * @param {Profile} profile - the profile to decorate.
     * @param {string[]} fields - the fields to include.
     * @param {string[]} expandFields - fields to expand.
     * @return {Profile} return the decorated profile.
     */
    ShippingProfileService.prototype.decorate = function (profile, fields, expandFields) {
        if (expandFields === void 0) { expandFields = []; }
        return __awaiter(this, void 0, void 0, function () {
            var requiredFields, decorated, _a, _b, final;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        requiredFields = ["_id", "metadata"];
                        decorated = lodash_1.default.pick(profile, fields.concat(requiredFields));
                        if (!(expandFields.includes("products") && profile.products)) return [3 /*break*/, 2];
                        _a = decorated;
                        return [4 /*yield*/, Promise.all(profile.products.map(function (pId) { return _this.productService_.retrieve(pId); }))];
                    case 1:
                        _a.products = _c.sent();
                        _c.label = 2;
                    case 2:
                        if (!(expandFields.includes("shipping_options") && profile.shipping_options)) return [3 /*break*/, 4];
                        _b = decorated;
                        return [4 /*yield*/, Promise.all(profile.shipping_options.map(function (oId) {
                                return _this.shippingOptionService_.retrieve(oId);
                            }))];
                    case 3:
                        _b.shipping_options = _c.sent();
                        _c.label = 4;
                    case 4: return [4 /*yield*/, this.runDecorators_(decorated)];
                    case 5:
                        final = _c.sent();
                        return [2 /*return*/, final];
                }
            });
        });
    };
    /**
     * Returns a list of all the productIds in the cart.
     * @param {Cart} cart - the cart to extract products from
     * @return {[string]} a list of product ids
     */
    ShippingProfileService.prototype.getProfilesInCart_ = function (cart) {
        return cart.items.reduce(function (acc, next) {
            // We may have line items that are not associated with a product
            if (next.variant && next.variant.product) {
                if (!acc.includes(next.variant.product.profile_id)) {
                    acc.push(next.variant.product.profile_id);
                }
            }
            return acc;
        }, []);
    };
    /**
     * Finds all the shipping profiles that cover the products in a cart, and
     * validates all options that are available for the cart.
     * @param {Cart} cart - the cart object to find shipping options for
     * @return {[ShippingOption]} a list of the available shipping options
     */
    ShippingProfileService.prototype.fetchCartOptions = function (cart) {
        return __awaiter(this, void 0, void 0, function () {
            var profileIds, selector, customShippingOptions, hasCustomShippingOptions, rawOpts, options, rawOpts_1, rawOpts_1_1, o, option;
            var e_4, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        profileIds = this.getProfilesInCart_(cart);
                        selector = {
                            profile_id: profileIds,
                            admin_only: false,
                        };
                        return [4 /*yield*/, this.customShippingOptionService_.list({
                                cart_id: cart.id,
                            }, { select: ["id", "shipping_option_id", "price"] })];
                    case 1:
                        customShippingOptions = _b.sent();
                        hasCustomShippingOptions = customShippingOptions === null || customShippingOptions === void 0 ? void 0 : customShippingOptions.length;
                        // if there are custom shipping options associated with the cart, use those
                        if (hasCustomShippingOptions) {
                            selector.id = customShippingOptions.map(function (cso) { return cso.shipping_option_id; });
                        }
                        return [4 /*yield*/, this.shippingOptionService_.list(selector, {
                                relations: ["requirements", "profile"],
                            })
                            // if there are custom shipping options associated with the cart, return cart shipping options with custom price
                        ];
                    case 2:
                        rawOpts = _b.sent();
                        // if there are custom shipping options associated with the cart, return cart shipping options with custom price
                        if (hasCustomShippingOptions) {
                            return [2 /*return*/, rawOpts.map(function (so) {
                                    var customOption = customShippingOptions.find(function (cso) { return cso.shipping_option_id === so.id; });
                                    return __assign(__assign({}, so), { amount: customOption === null || customOption === void 0 ? void 0 : customOption.price });
                                })];
                        }
                        options = [];
                        try {
                            for (rawOpts_1 = __values(rawOpts), rawOpts_1_1 = rawOpts_1.next(); !rawOpts_1_1.done; rawOpts_1_1 = rawOpts_1.next()) {
                                o = rawOpts_1_1.value;
                                try {
                                    option = this.shippingOptionService_.validateCartOption(o, cart);
                                    if (option) {
                                        options.push(option);
                                    }
                                }
                                catch (ex) {
                                    // catch the error, but intentionally do not break the iterations
                                }
                            }
                        }
                        catch (e_4_1) { e_4 = { error: e_4_1 }; }
                        finally {
                            try {
                                if (rawOpts_1_1 && !rawOpts_1_1.done && (_a = rawOpts_1.return)) _a.call(rawOpts_1);
                            }
                            finally { if (e_4) throw e_4.error; }
                        }
                        return [2 /*return*/, options];
                }
            });
        });
    };
    return ShippingProfileService;
}(medusa_interfaces_1.BaseService));
exports.default = ShippingProfileService;
