// @medusajs/medusa/dist/services/store.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var medusa_core_utils_1 = require("medusa-core-utils");
var medusa_interfaces_1 = require("medusa-interfaces");
var currencies_1 = require("../utils/currencies");
/**
 * Provides layer to manipulate store settings.
 * @extends BaseService
 */
var StoreService = /** @class */ (function (_super) {
    __extends(StoreService, _super);
    function StoreService(_a) {
        var manager = _a.manager, storeRepository = _a.storeRepository, currencyRepository = _a.currencyRepository, eventBusService = _a.eventBusService;
        var _this = _super.call(this) || this;
        /** @private @const {EntityManager} */
        _this.manager_ = manager;
        /** @private @const {StoreRepository} */
        _this.storeRepository_ = storeRepository;
        /** @private @const {CurrencyRepository} */
        _this.currencyRepository_ = currencyRepository;
        /** @private @const {EventBus} */
        _this.eventBus_ = eventBusService;
        return _this;
    }
    StoreService.prototype.withTransaction = function (transactionManager) {
        if (!transactionManager) {
            return this;
        }
        var cloned = new StoreService({
            manager: transactionManager,
            storeRepository: this.storeRepository_,
            currencyRepository: this.currencyRepository_,
            eventBusService: this.eventBus_,
        });
        cloned.transactionManager_ = transactionManager;
        return cloned;
    };
    /**
     * Creates a store if it doesn't already exist.
     * @return {Promise<Store>} the store.
     */
    StoreService.prototype.create = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var storeRepository, store, s;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    storeRepository = manager.getCustomRepository(this.storeRepository_);
                                    return [4 /*yield*/, this.retrieve()];
                                case 1:
                                    store = _a.sent();
                                    if (!!store) return [3 /*break*/, 4];
                                    return [4 /*yield*/, storeRepository.create()];
                                case 2:
                                    s = _a.sent();
                                    return [4 /*yield*/, storeRepository.save(s)];
                                case 3:
                                    store = _a.sent();
                                    _a.label = 4;
                                case 4: return [2 /*return*/, store];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Retrieve the store settings. There is always a maximum of one store.
     * @param {string[]} relations - relations to fetch with store
     * @return {Promise<Store>} the store
     */
    StoreService.prototype.retrieve = function (relations) {
        if (relations === void 0) { relations = []; }
        return __awaiter(this, void 0, void 0, function () {
            var storeRepo, store;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        storeRepo = this.manager_.getCustomRepository(this.storeRepository_);
                        return [4 /*yield*/, storeRepo.findOne({ relations: relations })];
                    case 1:
                        store = _a.sent();
                        return [2 /*return*/, store];
                }
            });
        });
    };
    StoreService.prototype.getDefaultCurrency_ = function (code) {
        var currencyObject = currencies_1.currencies[code.toUpperCase()];
        return {
            code: currencyObject.code.toLowerCase(),
            symbol: currencyObject.symbol,
            symbol_native: currencyObject.symbol_native,
            name: currencyObject.name,
        };
    };
    /**
     * Updates a store
     * @param {object} update - an object with the update values.
     * @return {Promise} resolves to the update result.
     */
    StoreService.prototype.update = function (update) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var storeRepository, currencyRepository, store, metadata, default_currency_code, storeCurrencies, rest, curr, _a, _b, _c, _d, key, value, result;
                        var e_1, _e;
                        var _this = this;
                        return __generator(this, function (_f) {
                            switch (_f.label) {
                                case 0:
                                    storeRepository = manager.getCustomRepository(this.storeRepository_);
                                    currencyRepository = manager.getCustomRepository(this.currencyRepository_);
                                    return [4 /*yield*/, this.retrieve()];
                                case 1:
                                    store = _f.sent();
                                    metadata = update.metadata, default_currency_code = update.default_currency_code, storeCurrencies = update.currencies, rest = __rest(update, ["metadata", "default_currency_code", "currencies"]);
                                    if (metadata) {
                                        store.metadata = this.setMetadata_(store.id, metadata);
                                    }
                                    if (!default_currency_code) return [3 /*break*/, 3];
                                    return [4 /*yield*/, currencyRepository.findOne({
                                            code: default_currency_code.toLowerCase(),
                                        })];
                                case 2:
                                    curr = _f.sent();
                                    if (!curr) {
                                        throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "Currency ".concat(default_currency_code, " not found"));
                                    }
                                    store.default_currency = curr;
                                    store.default_currency_code = curr.code;
                                    _f.label = 3;
                                case 3:
                                    if (!storeCurrencies) return [3 /*break*/, 5];
                                    _a = store;
                                    return [4 /*yield*/, Promise.all(storeCurrencies.map(function (curr) { return __awaiter(_this, void 0, void 0, function () {
                                            var currency;
                                            return __generator(this, function (_a) {
                                                switch (_a.label) {
                                                    case 0: return [4 /*yield*/, currencyRepository.findOne({
                                                            where: { code: curr.toLowerCase() },
                                                        })];
                                                    case 1:
                                                        currency = _a.sent();
                                                        if (!currency) {
                                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "Invalid currency ".concat(curr));
                                                        }
                                                        return [2 /*return*/, currency];
                                                }
                                            });
                                        }); }))];
                                case 4:
                                    _a.currencies = _f.sent();
                                    _f.label = 5;
                                case 5:
                                    try {
                                        for (_b = __values(Object.entries(rest)), _c = _b.next(); !_c.done; _c = _b.next()) {
                                            _d = __read(_c.value, 2), key = _d[0], value = _d[1];
                                            store[key] = value;
                                        }
                                    }
                                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                                    finally {
                                        try {
                                            if (_c && !_c.done && (_e = _b.return)) _e.call(_b);
                                        }
                                        finally { if (e_1) throw e_1.error; }
                                    }
                                    return [4 /*yield*/, storeRepository.save(store)];
                                case 6:
                                    result = _f.sent();
                                    return [2 /*return*/, result];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Add a currency to the store
     * @param {string} code - 3 character ISO currency code
     * @return {Promise} result after update
     */
    StoreService.prototype.addCurrency = function (code) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var storeRepo, currencyRepository, store, curr, updated;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    storeRepo = manager.getCustomRepository(this.storeRepository_);
                                    currencyRepository = manager.getCustomRepository(this.currencyRepository_);
                                    return [4 /*yield*/, this.retrieve(["currencies"])];
                                case 1:
                                    store = _a.sent();
                                    return [4 /*yield*/, currencyRepository.findOne({
                                            where: { code: code.toLowerCase() },
                                        })];
                                case 2:
                                    curr = _a.sent();
                                    if (!curr) {
                                        throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "Currency ".concat(code, " not found"));
                                    }
                                    if (store.currencies.map(function (c) { return c.code; }).includes(curr.code.toLowerCase())) {
                                        throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.DUPLICATE_ERROR, "Currency already added");
                                    }
                                    store.currencies = __spreadArray(__spreadArray([], __read(store.currencies), false), [curr], false);
                                    return [4 /*yield*/, storeRepo.save(store)];
                                case 3:
                                    updated = _a.sent();
                                    return [2 /*return*/, updated];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Removes a currency from the store
     * @param {string} code - 3 character ISO currency code
     * @return {Promise} result after update
     */
    StoreService.prototype.removeCurrency = function (code) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (manager) { return __awaiter(_this, void 0, void 0, function () {
                        var storeRepo, store, exists, updated;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    storeRepo = manager.getCustomRepository(this.storeRepository_);
                                    return [4 /*yield*/, this.retrieve(["currencies"])];
                                case 1:
                                    store = _a.sent();
                                    exists = store.currencies.find(function (c) { return c.code === code.toLowerCase(); });
                                    // If currency does not exist, return early
                                    if (!exists) {
                                        return [2 /*return*/, store];
                                    }
                                    store.currencies = store.currencies.filter(function (c) { return c.code !== code; });
                                    return [4 /*yield*/, storeRepo.save(store)];
                                case 2:
                                    updated = _a.sent();
                                    return [2 /*return*/, updated];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Decorates a store object.
     * @param {Store} store - the store to decorate.
     * @param {string[]} fields - the fields to include.
     * @param {string[]} expandFields - fields to expand.
     * @return {Store} return the decorated Store.
     */
    StoreService.prototype.decorate = function (store, fields, expandFields) {
        if (expandFields === void 0) { expandFields = []; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, store];
            });
        });
    };
    return StoreService;
}(medusa_interfaces_1.BaseService));
exports.default = StoreService;
