// @medusajs/medusa-js/dist/resources/admin/collections.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminCollectionsResource = /** @class */ (function (_super) {
    __extends(AdminCollectionsResource, _super);
    function AdminCollectionsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @description Creates a collection.
     * @param payload
     * @returns Created collection.
     */
    AdminCollectionsResource.prototype.create = function (payload) {
        var path = "/admin/collections";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description Updates a collection
     * @param id id of the collection to update.
     * @param payload update to apply to collection.
     * @returns the updated collection.
     */
    AdminCollectionsResource.prototype.update = function (id, payload) {
        var path = "/admin/collections/" + id;
        return this.client.request("POST", path, payload);
    };
    /**
     * @description deletes a collection
     * @param id id of collection to delete.
     * @returns Deleted response
     */
    AdminCollectionsResource.prototype.delete = function (id) {
        var path = "/admin/collections/" + id;
        return this.client.request("DELETE", path);
    };
    /**
     * @description get a collection
     * @param id id of the collection to retrieve.
     * @returns the collection with the given id
     */
    AdminCollectionsResource.prototype.retrieve = function (id) {
        var path = "/admin/collections/" + id;
        return this.client.request("GET", path);
    };
    /**
     * @description Lists collections matching a query
     * @param query Query for searching collections
     * @returns a list of collections matching the query.
     */
    AdminCollectionsResource.prototype.list = function (query) {
        var path = "/admin/collections";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/collections?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    return AdminCollectionsResource;
}(base_1.default));
exports.default = AdminCollectionsResource;
