// @medusajs/medusa-js/dist/resources/admin/gift-cards.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminGiftCardsResource = /** @class */ (function (_super) {
    __extends(AdminGiftCardsResource, _super);
    function AdminGiftCardsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @description Creates a gift card
     */
    AdminGiftCardsResource.prototype.create = function (payload) {
        var path = "/admin/gift-cards";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description Updates a gift card
     */
    AdminGiftCardsResource.prototype.update = function (id, payload) {
        var path = "/admin/gift-cards/" + id;
        return this.client.request("POST", path, payload);
    };
    /**
     * @description Deletes a gift card
     */
    AdminGiftCardsResource.prototype.delete = function (id) {
        var path = "/admin/gift-cards/" + id;
        return this.client.request("DELETE", path);
    };
    /**
     * @description Deletes a gift card
     */
    AdminGiftCardsResource.prototype.retrieve = function (id) {
        var path = "/admin/gift-cards/" + id;
        return this.client.request("GET", path);
    };
    /**
     * @description Lists gift cards
     */
    AdminGiftCardsResource.prototype.list = function (query) {
        var path = "/admin/gift-cards/";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/gift-cards?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    return AdminGiftCardsResource;
}(base_1.default));
exports.default = AdminGiftCardsResource;
