// @medusajs/medusa-js/dist/resources/admin/notes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminNotesResource = /** @class */ (function (_super) {
    __extends(AdminNotesResource, _super);
    function AdminNotesResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminNotesResource.prototype.create = function (payload) {
        var path = "/admin/notes";
        return this.client.request("POST", path, payload);
    };
    AdminNotesResource.prototype.update = function (id, payload) {
        var path = "/admin/notes/" + id;
        return this.client.request("POST", path, payload);
    };
    AdminNotesResource.prototype.delete = function (id) {
        var path = "/admin/notes/" + id;
        return this.client.request("DELETE", path);
    };
    AdminNotesResource.prototype.retrieve = function (id) {
        var path = "/admin/notes/" + id;
        return this.client.request("GET", path);
    };
    AdminNotesResource.prototype.list = function (query) {
        var path = "/admin/notes/";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return key + "=" + value;
            });
            path = "/admin/notes?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    return AdminNotesResource;
}(base_1.default));
exports.default = AdminNotesResource;
