// @medusajs/medusa-js/dist/resources/admin/orders.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminOrdersResource = /** @class */ (function (_super) {
    __extends(AdminOrdersResource, _super);
    function AdminOrdersResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminOrdersResource.prototype.create = function (payload) {
        var path = "/admin/orders";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.update = function (id, payload) {
        var path = "/admin/orders/" + id;
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.retrieve = function (id) {
        var path = "/admin/orders/" + id;
        return this.client.request("GET", path);
    };
    AdminOrdersResource.prototype.list = function (query) {
        var path = "/admin/orders";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return typeof value !== "undefined" ? key + "=" + value : "";
            });
            path = "/admin/orders?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    AdminOrdersResource.prototype.complete = function (id) {
        var path = "/admin/orders/" + id + "/complete";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.capturePayment = function (id) {
        var path = "/admin/orders/" + id + "/capture";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.refundPayment = function (id, payload) {
        var path = "/admin/orders/" + id + "/refund";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.createFulfillment = function (id, payload) {
        var path = "/admin/orders/" + id + "/fulfillment";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.cancelFulfillment = function (id, fulfillmentId) {
        var path = "/admin/orders/" + id + "/fulfillments/" + fulfillmentId + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.cancelSwapFulfillment = function (id, swapId, fulfillmentId) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/fulfillments/" + fulfillmentId + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.cancelClaimFulfillment = function (id, claimId, fulfillmentId) {
        var path = "/admin/orders/" + id + "/claims/" + claimId + "/fulfillments/" + fulfillmentId + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.createShipment = function (id, payload) {
        var path = "/admin/orders/" + id + "/shipment";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.requestReturn = function (id, payload) {
        var path = "/admin/orders/" + id + "/return";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.cancel = function (id) {
        var path = "/admin/orders/" + id + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.addShippingMethod = function (id, payload) {
        var path = "/admin/orders/" + id + "/shipping-methods";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.archive = function (id) {
        var path = "/admin/orders/" + id + "/archive";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.createSwap = function (id, payload) {
        var path = "/admin/orders/" + id + "/swaps";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.cancelSwap = function (id, swapId) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.receiveSwap = function (id, swapId, payload) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/receive";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.fulfillSwap = function (id, swapId, payload) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/fulfillments";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.createSwapShipment = function (id, swapId, payload) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/shipments";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.processSwapPayment = function (id, swapId) {
        var path = "/admin/orders/" + id + "/swaps/" + swapId + "/process-payment";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.createClaim = function (id, payload) {
        var path = "/admin/orders/" + id + "/claims";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.cancelClaim = function (id, claimId) {
        var path = "/admin/orders/" + id + "/claims/" + claimId + "/cancel";
        return this.client.request("POST", path);
    };
    AdminOrdersResource.prototype.updateClaim = function (id, claimId, payload) {
        var path = "/admin/orders/" + id + "/claims/" + claimId;
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.fulfillClaim = function (id, claimId, payload) {
        var path = "/admin/orders/" + id + "/claims/" + claimId + "/fulfillments";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.createClaimShipment = function (id, claimId, payload) {
        var path = "/admin/orders/" + id + "/claims/" + claimId + "/shipments";
        return this.client.request("POST", path, payload);
    };
    AdminOrdersResource.prototype.deleteMetadata = function (id, key) {
        var path = "/admin/orders/" + id + "/metadata/" + key;
        return this.client.request("DELETE", path);
    };
    return AdminOrdersResource;
}(base_1.default));
exports.default = AdminOrdersResource;
