// @medusajs/medusa-js/dist/resources/admin/products.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminProductsResource = /** @class */ (function (_super) {
    __extends(AdminProductsResource, _super);
    function AdminProductsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminProductsResource.prototype.create = function (payload) {
        var path = "/admin/products/";
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.retrieve = function (id) {
        var path = "/admin/products/" + id;
        return this.client.request("GET", path);
    };
    AdminProductsResource.prototype.update = function (id, payload) {
        var path = "/admin/products/" + id;
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.delete = function (id) {
        var path = "/admin/products/" + id;
        return this.client.request("DELETE", path);
    };
    AdminProductsResource.prototype.list = function (query) {
        var path = "/admin/products";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return typeof value !== "undefined" ? key + "=" + value : "";
            });
            path = "/admin/products?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    AdminProductsResource.prototype.listTypes = function () {
        var path = "/admin/products/types";
        return this.client.request("GET", path);
    };
    AdminProductsResource.prototype.listTags = function () {
        var path = "/admin/products/tag-usage";
        return this.client.request("GET", path);
    };
    AdminProductsResource.prototype.setMetadata = function (id, payload) {
        var path = "/admin/products/" + id + "/metadata";
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.createVariant = function (id, payload) {
        var path = "/admin/products/" + id + "/variants";
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.updateVariant = function (id, variantId, payload) {
        var path = "/admin/products/" + id + "/variants/" + variantId;
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.deleteVariant = function (id, variantId) {
        var path = "/admin/products/" + id + "/variants/" + variantId;
        return this.client.request("DELETE", path);
    };
    AdminProductsResource.prototype.addOption = function (id, payload) {
        var path = "/admin/products/" + id + "/options";
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.updateOption = function (id, optionId, payload) {
        var path = "/admin/products/" + id + "/options/" + optionId;
        return this.client.request("POST", path, payload);
    };
    AdminProductsResource.prototype.deleteOption = function (id, optionId) {
        var path = "/admin/products/" + id + "/options/" + optionId;
        return this.client.request("DELETE", path);
    };
    return AdminProductsResource;
}(base_1.default));
exports.default = AdminProductsResource;
