// @medusajs/medusa-js/dist/resources/admin/regions.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminRegionsResource = /** @class */ (function (_super) {
    __extends(AdminRegionsResource, _super);
    function AdminRegionsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @description creates a region.
     * @param payload
     * @returns created region.
     */
    AdminRegionsResource.prototype.create = function (payload) {
        var path = "/admin/regions";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description updates a region
     * @param id id of the region to update.
     * @param payload update to apply to region.
     * @returns the updated region.
     */
    AdminRegionsResource.prototype.update = function (id, payload) {
        var path = "/admin/regions/" + id;
        return this.client.request("POST", path, payload);
    };
    /**
     * @description deletes a region
     * @param id id of region to delete.
     * @returns Deleted response
     */
    AdminRegionsResource.prototype.delete = function (id) {
        var path = "/admin/regions/" + id;
        return this.client.request("DELETE", path);
    };
    /**
     * @description get a region
     * @param id id of the region to retrieve.
     * @returns the region with the given id
     */
    AdminRegionsResource.prototype.retrieve = function (id) {
        var path = "/admin/regions/" + id;
        return this.client.request("GET", path);
    };
    /**
     * @description lists regions matching a query
     * @param query query for searching regions
     * @returns a list of regions matching the query.
     */
    AdminRegionsResource.prototype.list = function (query) {
        var path = "/admin/regions";
        if (query) {
            var queryString = Object.entries(query).map(function (_a) {
                var key = _a[0], value = _a[1];
                return typeof value !== "undefined" ? key + "=" + value : "";
            });
            path = "/admin/regions?" + queryString.join("&");
        }
        return this.client.request("GET", path);
    };
    /**
     * @description adds metadata to a region
     * @param id region id
     * @param payload metadata
     * @returns updated region
     */
    AdminRegionsResource.prototype.setMetadata = function (id, payload) {
        var path = "/admin/regions/" + id + "/metadata";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description delete a region's metadata key value pair
     * @param id region id
     * @param key metadata key
     * @returns updated region
     */
    AdminRegionsResource.prototype.deleteMetadata = function (id, key) {
        var path = "/admin/regions/" + id + "/metadata/" + key;
        return this.client.request("DELETE", path);
    };
    /**
     * @description adds a country to the list of countries in a region
     * @param id region id
     * @param payload country data
     * @returns updated region
     */
    AdminRegionsResource.prototype.addCountry = function (id, payload) {
        var path = "/admin/regions/" + id + "/countries";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description remove a country from a region's list of coutnries
     * @param id region id
     * @param country_code the 2 character ISO code for the Country.
     * @returns updated region
     */
    AdminRegionsResource.prototype.deleteCountry = function (id, country_code) {
        var path = "/admin/regions/" + id + "/countries/" + country_code;
        return this.client.request("DELETE", path);
    };
    /**
     * @description adds a fulfillment provider to a region
     * @param id region id
     * @param payload fulfillment provider data
     * @returns updated region
     */
    AdminRegionsResource.prototype.addFulfillmentProvider = function (id, payload) {
        var path = "/admin/regions/" + id + "/fulfillment-providers";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description remove a fulfillment provider from a region
     * @param id region id
     * @param provider_id the if of the fulfillment provider
     * @returns updated region
     */
    AdminRegionsResource.prototype.deleteFulfillmentProvider = function (id, provider_id) {
        var path = "/admin/regions/" + id + "/fulfillment-providers/" + provider_id;
        return this.client.request("DELETE", path);
    };
    /**
     * @description retrieves the list of fulfillment options available in a region
     * @param id region id
     * @returns list of fulfillment options
     */
    AdminRegionsResource.prototype.retrieveFulfillmentOptions = function (id) {
        var path = "/admin/regions/" + id + "/fulfillment-options";
        return this.client.request("GET", path);
    };
    /**
     * @description adds a payment provider to a region
     * @param id region id
     * @param payload payment provider data
     * @returns updated region
     */
    AdminRegionsResource.prototype.addPaymentProvider = function (id, payload) {
        var path = "/admin/regions/" + id + "/payment-providers";
        return this.client.request("POST", path, payload);
    };
    /**
     * @description removes a payment provider from a region
     * @param id region id
     * @param provider_id the id of the payment provider
     * @returns updated region
     */
    AdminRegionsResource.prototype.deletePaymentProvider = function (id, provider_id) {
        var path = "/admin/regions/" + id + "/payment-providers/" + provider_id;
        return this.client.request("DELETE", path);
    };
    return AdminRegionsResource;
}(base_1.default));
exports.default = AdminRegionsResource;
