// @medusajs/medusa-js/dist/resources/admin/shipping-profiles.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminShippingProfilesResource = /** @class */ (function (_super) {
    __extends(AdminShippingProfilesResource, _super);
    function AdminShippingProfilesResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminShippingProfilesResource.prototype.create = function (payload) {
        var path = "/admin/shipping-profiles/";
        return this.client.request("POST", path, payload);
    };
    AdminShippingProfilesResource.prototype.update = function (id, payload) {
        var path = "/admin/shipping-profiles/" + id;
        return this.client.request("POST", path, payload);
    };
    AdminShippingProfilesResource.prototype.delete = function (id) {
        var path = "/admin/shipping-profiles/" + id;
        return this.client.request("DELETE", path);
    };
    AdminShippingProfilesResource.prototype.retrieve = function (id) {
        var path = "/admin/shipping-profiles/" + id;
        return this.client.request("GET", path);
    };
    AdminShippingProfilesResource.prototype.list = function () {
        var path = "/admin/shipping-profiles/";
        return this.client.request("GET", path);
    };
    return AdminShippingProfilesResource;
}(base_1.default));
exports.default = AdminShippingProfilesResource;
