// @medusajs/medusa/dist/helpers/test-request.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.request = exports.clientSessionOpts = exports.adminSessionOpts = void 0;
var awilix_1 = require("awilix");
var express_1 = __importDefault(require("express"));
var jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
var medusa_test_utils_1 = require("medusa-test-utils");
require("reflect-metadata");
var supertest_1 = __importDefault(require("supertest"));
var config_1 = __importDefault(require("../config"));
var api_1 = __importDefault(require("../loaders/api"));
var passport_1 = __importDefault(require("../loaders/passport"));
var services_1 = __importDefault(require("../loaders/services"));
var adminSessionOpts = {
    cookieName: "session",
    secret: "test",
};
exports.adminSessionOpts = adminSessionOpts;
var clientSessionOpts = {
    cookieName: "session",
    secret: "test",
};
exports.clientSessionOpts = clientSessionOpts;
var testApp = (0, express_1.default)();
var container = (0, awilix_1.createContainer)();
container.register({
    logger: (0, awilix_1.asValue)({
        error: function () { },
    }),
    manager: (0, awilix_1.asValue)(medusa_test_utils_1.MockManager),
});
testApp.set("trust proxy", 1);
testApp.use(function (req, res, next) {
    req.session = {};
    var data = req.get("Cookie");
    if (data) {
        req.session = __assign(__assign({}, req.session), JSON.parse(data));
    }
    next();
});
(0, services_1.default)({ container: container });
(0, passport_1.default)({ app: testApp, container: container });
testApp.use(function (req, res, next) {
    req.scope = container.createScope();
    next();
});
(0, api_1.default)({ container: container, rootDirectory: ".", app: testApp });
var supertestRequest = (0, supertest_1.default)(testApp);
function request(method, url, opts) {
    if (opts === void 0) { opts = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var payload, headers, req, name, res, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    payload = opts.payload, headers = opts.headers;
                    req = supertestRequest[method.toLowerCase()](url);
                    headers = headers || {};
                    headers.Cookie = headers.Cookie || "";
                    if (opts.adminSession) {
                        if (opts.adminSession.jwt) {
                            opts.adminSession.jwt = jsonwebtoken_1.default.sign(opts.adminSession.jwt, config_1.default.jwtSecret, {
                                expiresIn: "30m",
                            });
                        }
                        headers.Cookie = JSON.stringify(opts.adminSession) || "";
                    }
                    if (opts.clientSession) {
                        if (opts.clientSession.jwt) {
                            opts.clientSession.jwt = jsonwebtoken_1.default.sign(opts.clientSession.jwt, config_1.default.jwtSecret, {
                                expiresIn: "30d",
                            });
                        }
                        headers.Cookie = JSON.stringify(opts.clientSession) || "";
                    }
                    for (name in headers) {
                        req.set(name, headers[name]);
                    }
                    if (payload && !req.get("content-type")) {
                        req.set("Content-Type", "application/json");
                    }
                    if (!req.get("accept")) {
                        req.set("Accept", "application/json");
                    }
                    req.set("Host", "localhost");
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, req.send(JSON.stringify(payload))];
                case 2:
                    res = _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _a.sent();
                    if (e_1.response) {
                        res = e_1.response;
                    }
                    else {
                        throw e_1;
                    }
                    return [3 /*break*/, 4];
                case 4: 
                // let c =
                //  res.headers["set-cookie"] && cookie.parse(res.headers["set-cookie"][0])
                // res.adminSession =
                //  c &&
                //  c[adminSessionOpts.cookieName] &&
                //  sessions.util.decode(adminSessionOpts, c[adminSessionOpts.cookieName])
                //    .content
                // res.clientSession =
                //  c &&
                //  c[clientSessionOpts.cookieName] &&
                //  sessions.util.decode(clientSessionOpts, c[clientSessionOpts.cookieName])
                //    .content
                return [2 /*return*/, res];
            }
        });
    });
}
exports.request = request;
